src/python/pants/backend/shell/target_types.py

```diff
@@ -109,19 +110,45 @@ def compute_value(cls, raw_value: Optional[int], address: Address) -> Optional[i
         return value
 
 
+class Shunit2TestSourceField(ShellSourceField):
+    pass
+
+
 class Shunit2ShellField(StringField):
     alias = "shell"
     valid_choices = Shunit2Shell
     help = "Which shell to run the tests with. If unspecified, Pants will look for a shebang line."
 
 
-class Shunit2Tests(Target):
+class Shunit2TestTarget(Target):
+    alias = "shunit2_tests"  # TODO(#12954): rename to `shunit_test` when ready. Update `help` too.
+    core_fields = (
+        *COMMON_TARGET_FIELDS,
+        Shunit2TestSourceField,
+        Shunit2TestDependenciesField,
+        Shunit2TestTimeoutField,
+        Shunit2ShellField,
+        RuntimePackageDependenciesField,
+    )
+    help = "A test file for Bourne-based shell scripts using the shunit2 test framework."
+
+
+# -----------------------------------------------------------------------------------------------
+# `shunit2_tests` target generator
+# -----------------------------------------------------------------------------------------------
+
+
+class Shunit2TestsGeneratorSourcesField(ShellGeneratingSources):
+    default = ("*_test.sh", "test_*.sh", "tests.sh")
+
+
+class Shunit2TestsGeneratorTarget(Target):
```

Comment:
This name is verbose, but I like it for making abundantly clear its purpose. 

I thought about using `Macro` instead of `GeneratorTarget`, but that's a misnomer. An important part of https://docs.google.com/document/d/1HpJn2jTWf5sKob6zhe4SqHZ7KWBlX4a6OJOwnuw-IHo/edit#heading=h.7qrnzjjahn5v is realizing that target generators are targets themselves.